### ISSUE TITLE
Consider switching to Idiomorph

### ISSUE THREAD

**Issue by hirasso** [daun/swup-morph-plugin#14]
Morphdom has a few known limitations if it comes to id-less parent elements. That's why the creator of [HTMX](https://github.com/bigskysoftware/htmx) has developed an alternative library called [Idiomorph](https://github.com/bigskysoftware/idiomorph). The improvements are being presented quite nicely in the [library's readme](https://github.com/bigskysoftware/idiomorph#demo).

Maybe it would be worth a thought for morph plugin to consider switching to Idiodom?

**Comment by daun** [comment]
Thanks for calling my attention to idiomorph, it certainly looks interesting. I'll need to check it out and see about backward compatibility. The options and callbacks look like they follow their own naming conventions, unfortunately.

This might be a good fit for the head plugin as well? Not sure if it understands the order of style tags, but probably worth giving it a go.

**Comment by hirasso** [comment]
I haven't used head plugin yet, so I can't tell if Idiomorph would be a good fit for that plugin as well. But I think it's good to be aware that it exists and especially of the limitations of Morphdom

**Comment by daun** [comment]
Note to self: idiomorph still looks very interesting but currently has two major flaws: it doesn't support returning `false` from callbacks to prevent replacing or removing nodes, and it's distributed as AMD only. There's PRs for both of these in the pipeline, but nothing released yet.

**Comment by daun** [comment]
I wonder if [morphlex](https://github.com/yippee-fun/morphlex) isn't the right choice nowadays. It is smaller, has tons of useful callbacks for controlling the morph, and incorporates many morphing optimizations like the id matching algorithm introduced by idiomorph. Both are really well tested.

**Comment by hirasso** [comment]
Ha, I also stumbled upon morphlex a few days ago 🤠 

I currently lack the time to look into it more deeply, too much client work... but either would be fine with me :)

**Comment by joeldrapper** [comment]
If you have any questions about Morphlex or run into any issues, do let me know. It’s a much less popular library but we’re using it in production at Plane to morph forms as you fill them out, and we haven’t had any issues yet. We have 100% line and branch coverage and run tests across Chromium, Firefox and Webkit.

**Comment by daun** [comment]
@joeldrapper Thanks for the nudge, I've [started work](https://github.com/daun/swup-morph-plugin/pull/41) on porting things over to Morphlex. One quick question:

We're currently checking `el.isEqualNode()` ourselves. That's no longer necessary when using Morphlex, right?

**Comment by hirasso** [comment]
> We're currently checking `el.isEqualNode()` ourselves. That's no longer necessary when using Morphlex, right?

Looks like it:

https://github.com/yippee-fun/morphlex/blob/bd0e7f8592523da0f1fabc9d65a4db1ad087d350/src/morphlex.ts#L351

If I understand correctly, morphlex' [`flagDirtyInputs()`](https://github.com/yippee-fun/morphlex/blob/bd0e7f8592523da0f1fabc9d65a4db1ad087d350/src/morphlex.ts#L212-L255) function also does a lot of the stuff that `swup-morph-plugin` needed to do manually before. Maybe a new major version would provide a good chance to further simplify the [`shouldMorphCallbacks`](https://github.com/daun/swup-morph-plugin/blob/28ada95c35c5d71eab9f605bb0deceef812d5c40/src/morph.ts#L82) callback functions?

**Comment by joeldrapper** [comment]
Yes, that seems likely. In order to make `isEqualNode` notice changed input values, we first iterate over inputs and set this dirty attribute on them if the input has changed from its default value. I explained the algorithm briefly here. https://joel.drapper.me/p/morphlex/

It has changed a little since I wrote this, but not much. I’m using slightly different data structures internally now for better performance.

**Comment by daun** [comment]
@hirasso Good idea, this is going to be a major version for sure. I think we can get rid of all of them except the permanent check, which should be converted to `data-swup-persist` now that swup supports that as well. We currently also have a special case for `contenteditable` which we could easily drop, it's not something I've ever actually used.

@joeldrapper That's a great writeup, really good explanation of the internals.

**Comment by hirasso** [comment]
@joeldrapper lovely writeup, indeed! The `preserveChanges` option looks exactly like the thing we should use for swup-morph-plugin.

@daun I actually didn't even know about the permanent check before reading the source today... by the way - why would someone want to persist an element when morphing? Have you had a use case for that feature in the past? In any case, if we decide to keep the feature, I agree that reusing `data-swup-persist` would make sense - and then we should document it in the README :)

**Comment by daun** [comment]
@hirasso The persist option was before swup supported it. I've used it quite a bit for making sure lazyloaded images that moved in the dom structure were still kept around. But yeah, we can remove it now for sure.

**Comment by daun** [comment]
Released as [v2](https://github.com/daun/swup-morph-plugin/releases/tag/2.0.0) 🤠
